No notebook abaixo eu busco resolver as etapas do desafio da Prodesp. Meu nome é Victor Barreto Mesquita.

# 1. Extração de dados

In [114]:
# Alguns pacotes necessários para a análise.
import requests as rq
import pprint as pp
import pandas as pd
import numpy as np

Abaixo eu defino os links que serão usados para extrair os dados do banco de dados do IBGE.

In [10]:
url_pib="https://servicodados.ibge.gov.br/api/v3/agregados/5938/periodos/-6/variaveis/37?localidades=N1[all]|N6[all]"

In [11]:
url_populacao_residente = "https://servicodados.ibge.gov.br/api/v3/agregados/793/periodos/2007/variaveis/93?localidades=N6[all]"

In [12]:
url_populacao_residente_estimada = "https://servicodados.ibge.gov.br/api/v3/agregados/6579/periodos/-6/variaveis/9324?localidades=N6[all]"

Nas linhas de código abaixo eu faço um requerimento para as url's definidas acima e obtenho uma resposta em forma de texto.

In [13]:
requisicao = rq.get(url_pib)

In [14]:
requisicao2 = rq.get(url_populacao_residente)

In [15]:
requisicao3 = rq.get(url_populacao_residente_estimada)

Abaixo eu obtenho o json do requerimento acima.

In [16]:
informacoes_pib = requisicao.json()

In [17]:
informacoes_populacao = requisicao2.json()

In [18]:
informacoes_populacao_residente_estimada = requisicao3.json()

Nas linhas de código abaixo eu realizo algumas operações nas listas e dicionários buscando obter as informações que eu quero

In [19]:
item_busca = informacoes_pib[0]['variavel']
resultados = informacoes_pib[0]['resultados'][0]['series']

In [20]:
item_busca2 = informacoes_populacao[0]['variavel']
resultados2 = informacoes_populacao[0]['resultados'][0]['series']

In [21]:
item_busca3 = informacoes_populacao_residente_estimada[0]['variavel']
resultados3 = informacoes_populacao_residente_estimada[0]['resultados'][0]['series']

Abaixo eu crio um loop para criar uma lista apenas com informações como id da cidade, nome da cidade e ano em que foi obtido os dados.

In [110]:
dados = []

In [111]:
for i in range(len(informacoes_pib[0]['resultados'][0]['series'])):
    dados.append(informacoes_pib[0]['resultados'][0]['series'][i]['localidade']['id'])
    dados.append(informacoes_pib[0]['resultados'][0]['series'][i]['localidade']['nome'])
    dados.append(informacoes_pib[0]['resultados'][0]['series'][i]['serie']['2021'])
    

Abaixo eu crio um dataframe com as informações acima.

In [115]:
matriz_np = np.array(dados)

In [118]:
matriz_ajustada = np.reshape(matriz_np,(len(informacoes_pib[0]['resultados'][0]['series']),3))

In [151]:
df_pib = pd.DataFrame(matriz_ajustada,columns=[['id','cidade','PIB']])
df_pib.reset_index(drop=True)
df_pib = df_pib[1:];df_pib

id                      cidade        PIB
1     1100015  Alta Floresta D'Oeste - RO     734469
2     1100023              Ariquemes - RO    3209761
3     1100031                 Cabixi - RO     238412
4     1100049                 Cacoal - RO    2792383
5     1100056             Cerejeiras - RO     743037
...       ...                         ...        ...
5566  5222005             Vianópolis - GO     857968
5567  5222054         Vicentinópolis - GO     480173
5568  5222203               Vila Boa - GO     179899
5569  5222302          Vila Propício - GO     308013
5570  5300108               Brasília - DF  286943782

[5570 rows x 3 columns]

Aqui eu repito os procedimentos acima para criar um dataframe com as informações sobre a população.

In [126]:
dados2=[]

In [127]:
for i in range(len(informacoes_populacao_residente_estimada[0]['resultados'][0]['series'])):
    dados2.append(informacoes_populacao_residente_estimada[0]['resultados'][0]['series'][i]['localidade']['id'])
    dados2.append(informacoes_populacao_residente_estimada[0]['resultados'][0]['series'][i]['localidade']['nome'])
    dados2.append(informacoes_populacao_residente_estimada[0]['resultados'][0]['series'][i]['serie']['2021'])

In [129]:
matriz_np2 = np.array(dados2)

In [130]:
matriz_ajustada2 = np.reshape(matriz_np2,(len(informacoes_populacao_residente_estimada[0]['resultados'][0]['series']),3))

In [327]:
df_populacao = pd.DataFrame(matriz_ajustada2,columns=[['id','cidade','populacao']])
df_populacao.index = df_pib.index

id                      cidade populacao
1     1100015  Alta Floresta D'Oeste - RO     22516
2     1100023              Ariquemes - RO    111148
3     1100031                 Cabixi - RO      5067
4     1100049                 Cacoal - RO     86416
5     1100056             Cerejeiras - RO     16088
...       ...                         ...       ...
5566  5222005             Vianópolis - GO     14088
5567  5222054         Vicentinópolis - GO      9002
5568  5222203               Vila Boa - GO      6451
5569  5222302          Vila Propício - GO      5941
5570  5300108               Brasília - DF   3094325

[5570 rows x 3 columns]

Abaixo eu acrescento o PIB municipal no dataframe de populações das cidades.

In [334]:
df_populacao['PIB'] = df_pib['PIB']
df_populacao['id']=df_populacao['id'].astype('int')
df_populacao.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5570 entries, 1 to 5570
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   (id,)         5570 non-null   int32 
 1   (cidade,)     5570 non-null   object
 2   (populacao,)  5570 non-null   object
 3   (PIB,)        5570 non-null   object
dtypes: int32(1), object(3)
memory usage: 152.4+ KB


Criei uma coluna chamada 'id2' para ficar mais parecida com o outro dataframe.

In [335]:
df_populacao['id2'] = df_populacao['id']//10

In [336]:
df_populacao

id                      cidade populacao        PIB     id2
1     1100015  Alta Floresta D'Oeste - RO     22516     734469  110001
2     1100023              Ariquemes - RO    111148    3209761  110002
3     1100031                 Cabixi - RO      5067     238412  110003
4     1100049                 Cacoal - RO     86416    2792383  110004
5     1100056             Cerejeiras - RO     16088     743037  110005
...       ...                         ...       ...        ...     ...
5566  5222005             Vianópolis - GO     14088     857968  522200
5567  5222054         Vicentinópolis - GO      9002     480173  522205
5568  5222203               Vila Boa - GO      6451     179899  522220
5569  5222302          Vila Propício - GO      5941     308013  522230
5570  5300108               Brasília - DF   3094325  286943782  530010

[5570 rows x 5 columns]

Eu não encontrei dados de IDH no site do IBGE, então busquei essa informação no site do IPEA e fiz o download do arquivo CSV.

In [304]:
IDH_data = pd.read_csv('C:\\Users\\User\\Downloads\\iDH_2010.csv',sep=';')
IDHM = IDH_data[['Código do Município','Município','Código da Unidade da Federação','Nome da Unidade da Federação','IDHM']]
IDHM.columns = ['id','cidade','codigo_uf','nome_uf','IDHM'];IDHM

id                 cidade  codigo_uf           nome_uf   IDHM
0     110001  ALTA FLORESTA D'OESTE         11          Rondônia  0,641
1     110002              ARIQUEMES         11          Rondônia  0,702
2     110003                 CABIXI         11          Rondônia   0,65
3     110004                 CACOAL         11          Rondônia  0,718
4     110005             CEREJEIRAS         11          Rondônia  0,692
...      ...                    ...        ...               ...    ...
5559  522200             VIANÓPOLIS         52             Goiás  0,712
5560  522205         VICENTINÓPOLIS         52             Goiás  0,684
5561  522220               VILA BOA         52             Goiás  0,647
5562  522230          VILA PROPÍCIO         52             Goiás  0,634
5563  530010               BRASÍLIA         53  Distrito Federal  0,824

[5564 rows x 5 columns]

Juntar as informações dessa base com a anterior pelo chave id

In [356]:
df_populacao.columns = ['id','cidade','populacao','PIB','id2']
df_populacao.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5570 entries, 1 to 5570
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         5570 non-null   int32 
 1   cidade     5570 non-null   object
 2   populacao  5570 non-null   object
 3   PIB        5570 non-null   object
 4   id2        5570 non-null   int32 
dtypes: int32(2), object(3)
memory usage: 174.2+ KB


In [360]:
df_final = pd.merge(df_populacao,IDHM,left_on='id2',right_on='id',how='inner')
df_final = df_final[['id2','cidade_x','populacao','PIB','codigo_uf','nome_uf','IDHM']]
df_final.columns = ['id','cidade','populacao','PIB','codigo_uf','nome_uf','IDHM']
df_final

id                      cidade populacao        PIB  codigo_uf  \
0     110001  Alta Floresta D'Oeste - RO     22516     734469         11   
1     110002              Ariquemes - RO    111148    3209761         11   
2     110003                 Cabixi - RO      5067     238412         11   
3     110004                 Cacoal - RO     86416    2792383         11   
4     110005             Cerejeiras - RO     16088     743037         11   
...      ...                         ...       ...        ...        ...   
5559  522200             Vianópolis - GO     14088     857968         52   
5560  522205         Vicentinópolis - GO      9002     480173         52   
5561  522220               Vila Boa - GO      6451     179899         52   
5562  522230          Vila Propício - GO      5941     308013         52   
5563  530010               Brasília - DF   3094325  286943782         53   

               nome_uf   IDHM  
0             Rondônia  0,641  
1             Rondônia  0,702  
2             Rondônia   0,65  
3             Rondônia  0,718  
4             Rondônia  0,692  
...                ...    ...  
5559             Goiás  0,712  
5560             Goiás  0,684  
5561             Goiás  0,647  
5562             Goiás  0,634  
5563  Distrito Federal  0,824  

[5564 rows x 7 columns]

# SNOWFLAKE

Nessa seção eu armazeno a base de dados criada acima em um banco de dados Snowflake.

In [362]:
dados_lst = df_final.values.tolist()

Definição dos parâmetros necessários para criar uma sessão do Snowflake.

In [363]:
from snowflake.snowpark import Session
connection_parameters = {
   "account": "iba52159.east-us-2.azure",
   "user": "civildorock",
   "password": "Daronsoad1993@",
   "role": "ACCOUNTADMIN",  # optional
   #"warehouse": "<your snowflake warehouse>",  # optional
   "database": "DESAFIO"  # optional
   ,"schema": "PUBLIC"  # optional

 }  

# Começo configurando uma sessão para o Snowflake.

session = Session.builder.configs(connection_parameters).create() 

# 2. Transformação de Dados com PySpark e SQL:

Aqui executo os códigos do PySpark usando Snowflake Spark.

In [368]:
from snowflake.snowpark.functions import col
# Criar o dataframe que extraimos no IBGE e IPEA e escrevê-lo dentro da cloud Snowflake
df = session.create_dataframe(dados_lst, schema=["ID", "cidade","populacao", "PIB",'codigo_uf','nome_uf','IDHM'])
df.write.mode("overwrite").save_as_table("desafio_data")
session.table("desafio_data").collect()
df.write.save_as_table("desafio_data", mode="overwrite")
session.table("desafio_data").collect()
df_table = session.table("desafio_data")
# Mostrar a tabela criada no snowflake
df_table.show()

-----------------------------------------------------------------------------------------------------------
|"ID"    |"CIDADE"                    |"POPULACAO"  |"PIB"      |"CODIGO_UF"  |"NOME_UF"         |"IDHM"  |
-----------------------------------------------------------------------------------------------------------
|110001  |Alta Floresta D'Oeste - RO  |22516        |734469     |11           |Rondônia          |0,641   |
|530010  |Brasília - DF               |3094325      |286943782  |53           |Distrito Federal  |0,824   |
|522230  |Vila Propício - GO          |5941         |308013     |52           |Goiás             |0,634   |
|522220  |Vila Boa - GO               |6451         |179899     |52           |Goiás             |0,647   |
|522205  |Vicentinópolis - GO         |9002         |480173     |52           |Goiás             |0,684   |
|522200  |Vianópolis - GO             |14088        |857968     |52           |Goiás             |0,712   |
|522190  |Varjão - GO       

Verificar se existem duplicados usando o SQL

In [378]:
duplicate = session.sql("select id, count(*) from desafio_data group by ID having count(*)>1")
duplicate.show()

---------------------
|"ID"  |"COUNT(*)"  |
---------------------
|      |            |
---------------------



Ver a quantidade de linhas

In [380]:
num_linhas = session.sql("select count(*) as contagem_linha from desafio_data")
num_linhas.show()

--------------------
|"CONTAGEM_LINHA"  |
--------------------
|5564              |
--------------------



# 3. Modelagem e Documentação

A tabela final possui 5564 registros informações de municípios Brasileiros como id do município, nome do município, sigla do estado, nome do estado,população, PIB e IDH. Os dados foram coletados no ano de 2021 no site do IBGE e do IPEA. 

Tive o cuidado de escrever todos os passos no notebook antes de cada linha de código objetivando ficar claro tudo que fiz.